src/basic/hashmap.c

```diff
@@ -2147,13 +2150,48 @@ int _hashmap_dump_sorted(HashmapBase *h, void ***ret, size_t *ret_n) {
         assert(n == _hashmap_size(h));
         entries[n] = NULL;
 
-        typesafe_qsort_r(entries, n, hashmap_entry_compare, h->hash_ops->compare);
+        typesafe_qsort_r((struct hashmap_base_entry**) entries, n,
+                         hashmap_entry_compare, h->hash_ops->compare);
+
+        *ret = TAKE_PTR(entries);
+        *ret_n = n;
+        return 0;
+}
+
+int _hashmap_dump_keys_sorted(HashmapBase *h, void ***ret, size_t *ret_n) {
+        _cleanup_free_ void **entries = NULL;
+        size_t n;
+        int r;
+
+        r = _hashmap_dump_entries_sorted(h, &entries, &n);
+        if (r < 0)
+                return r;
+
+        /* Reuse the array. */
+        FOREACH_ARRAY(e, entries, n)
+                *e = (void*) (*(struct hashmap_base_entry**) e)->key;
```

Comment:
Reworked the handling of reusing the array, as requested in https://github.com/systemd/systemd/pull/30840#discussion_r1484434252. Is this better

My question from there also still stands:
> since the compiler can assume that pointers are always aligned to the alignment of their type

I'm not quite sure I'm following this here. Pointers are always 8 bytes, and we're allocating the array via malloc. Arrays are always tightly packed in C, no? So what room does the compiler have to care about alignment?

Reply:
the compiler may assume that the lower bits of a pointer to a type with alignment X are all zero, because the alignment implies that. now, you previously used an entry of pointers to some struct, but then stored generic pointers in them which might have any alignment (and thus must be assumed to have alignment 1). So suddenly you stored pointers where every single bit matters into fields where you told the compiler first how many of the lowest bunch of bits don't matter. And that's not OK.

Now, I think IRL the code will just run fine. But the thing is that the undefined behaviour concept in C allows the compiler to optimize hell out of a pointer if it knows the alignment of it and thus can assume how many zero bits are in the lowest bits...